### DISCUSSION TITLE
Connection error with Miltiple Certificate

### DISCUSSION THREAD

**Discussion by drcyberg** [strongswan/strongswan#1822] [Q&A] [answered]
Hi @tobiasbrunner!

I want to create two connections with different settings. I Generating a two Certificate for the one VPN Server. `"setup.vpn.cp.net"` and `"vpn.cp.net"` and accordingly I created two subnet DNS record.
I can connect with `ca-cert.pem` to `vpn.cp.net`, BUT I can't connect to with `ca-setup-cert.pem` to `setup.vpn.cp.net` anymore. :(

**Setup VPN:** Remote Access with Radius service

Here is the settings and logs:

**ipsec.conf**
<details><summary>Details</summary>
<p>

config setup
    charondebug="ike 1, knl 1, cfg 0"
    uniqueids=no

conn %default
    ikelifetime=60m
    keylife=20m
    rekeymargin=3m
    keyingtries=1
    keyexchange=ikev2

conn cp-vpn
    left=xxx.xxx.xxx.xxx
    leftsendcert=always
    leftauth=pubkey
    leftfirewall=yes
    leftsourceip=%config
    rightsendcert=never
    rightauth=eap-radius
    mobike=no
    eap_identity=%identity
    leftauth=pubkey
    type=tunnel
    forceencaps=yes
    compress=no
    fragmentation=no
    ike=chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,aes256-aes128-sha256-sha1-modp3072-modp2048-modp1024,3des-sha1-modp1024
    esp=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1-ecp256,aes256-sha1,3des-sha1

conn cp-access-accept-vpn
    also=cp-vpn
    leftcert=server-cert.pem
    leftid=vpn.cyberpointer.net
    leftsubnet=0.0.0.0/0
    rightid=%any
    right=%any
    rightsourceip=10.10.10.0/30
    #rightsubnet=192.168.0.0/24,192.168.1.0/24
    rightdns=8.8.8.8,8.8.4.4
    auto=add  

conn cp-setup-accept-vpn
    also=cp-vpn
    leftcert=setup-server-cert.pem
    leftid=setup.vpn.cyberpointer.net
    leftsubnet=11.11.11.0/24
    rightid=%any
    right=%any
    rightsourceip=11.11.11.0/24
    rightsubnet=11.11.11.0/24
    auto=add
    inactivity=10

</p>
</details>

**ipsec.secrets**
<details><summary>Details</summary>
<p> 
# This file holds shared secrets or RSA private keys for authentication.

# RSA private key for this host, authenticating it to any other host
# which knows the public part.

vpn.cyberpointer.net : RSA server-key.pem
setup.vpn.cyberpointer.net : RSA server-setup-key.pem

</p>
</details>

**syslog - server site**
<details><summary>Details</summary>
<p>

...
Aug  1 03:42:31 vpn ipsec[11376]: 06[IKE] received cert request for "CN=CP VPN root CA"
Aug  1 03:42:31 vpn ipsec[11376]: 06[IKE] initiating EAP_IDENTITY method (id 0x00)
Aug  1 03:42:31 vpn ipsec[11376]: 06[IKE] peer supports MOBIKE, but disabled in config
Aug  1 03:42:31 vpn ipsec[11376]: 06[IKE] authentication of 'vpn.cyberpointer.net' (myself) with RSA_EMSA_PKCS1_SHA2_384 successful
Aug  1 03:42:31 vpn ipsec[11376]: 06[IKE] sending end entity cert "CN=vpn.cyberpointer.net"
Aug  1 03:42:31 vpn ipsec[11376]: 06[ENC] generating IKE_AUTH response 1 [ IDr CERT AUTH EAP/REQ/ID ]
Aug  1 03:42:31 vpn ipsec[11376]: 06[NET] sending packet: from xxx.xxx.xxx.xxx[4500] to 79.122.87.110[37386] (1954 bytes)
Aug  1 03:42:31 vpn ipsec[11376]: 11[NET] received packet: from 79.122.87.110[37386] to xxx.xxx.xxx.xxx[4500] (65 bytes)
Aug  1 03:42:31 vpn ipsec[11376]: 11[ENC] parsed INFORMATIONAL request 2 [ N(AUTH_FAILED) ]
Aug  1 03:42:31 vpn ipsec[11376]: 11[ENC] generating INFORMATIONAL response 2 [ N(AUTH_FAILED) ]
Aug  1 03:42:31 vpn charon: 11[NET] sending packet: from xxx.xxx.xxx.xxx[4500] to 79.122.87.110[37386] (65 bytes)

</p>
</details> 

**syslog - client site**
<details><summary>Details</summary>
<p>

...
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 08[NET] sending packet: from 192.168.1.152[37386] to 193.188.192.136[4500] (611 bytes)
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[NET] received packet: from xxx.xxx.xxx.xxx[4500] to 192.168.1.152[37386] (1954 bytes)
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[ENC] parsed IKE_AUTH response 1 [ IDr CERT AUTH EAP/REQ/ID ]
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[IKE] received end entity cert "CN=vpn.cyberpointer.net"
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[CFG]   using certificate "CN=vpn.cyberpointer.net"
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[CFG]   using trusted ca certificate "CN=CP VPN root CA"
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[CFG]   reached self-signed root ca with a path length of 0
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[CFG] checking certificate status of "CN=vpn.cyberpointer.net"
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[CFG] certificate status is not available
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[IKE] authentication of 'vpn.cyberpointer.net' with RSA_EMSA_PKCS1_SHA2_384 successful
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[CFG] constraint check failed: identity 'setup.vpn.cyberpointer.net' required 
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[CFG] selected peer config 'CP-VPN' unacceptable: constraint checking failed
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[CFG] no alternative config found
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[ENC] generating INFORMATIONAL request 2 [ N(AUTH_FAILED) ]
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 charon-nm: 07[NET] sending packet: from 192.168.1.152[37386] to 193.188.192.136[4500] (65 bytes)
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 NetworkManager[901]: <warn>  [1690861277.9276] vpn[0x555ed187e830,52fc3129-c5b5-4da8-a99f-dc262bc319a5,"CP-VPN"]: dbus: failure: connect-failed (1)
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 NetworkManager[901]: <warn>  [1690861277.9277] vpn[0x555ed187e830,52fc3129-c5b5-4da8-a99f-dc262bc319a5,"CP-VPN"]: dbus: failure: connect-failed (1)
Aug  1 05:41:17 drcyberg-Precision-Tower-5810 NetworkManager[901]: <warn>  [1690861277.9280] vpn[0x555ed187e830,52fc3129-c5b5-4da8-a99f-dc262bc319a5,"CP-VPN"]: dbus: failure: login-failed (0)

</p>
</details>

Is this not possible or should it be solved in a completely different way?
Or it only works with PKCS#11 Cryptographic Token?
Is not possible the subjectAltName (`-- san`) selector?

Thank you!

Regards: DrCyberg

**Comment by tobiasbrunner** [comment]
The NetworkManager plugin uses the configured hostname as identity, by default, but does not send it to the peer (i.e. IDr is not sent). So the server won't have an identity to match against the configs and will just use the first one.

To make the NM client explicitly send a remote identity, you have to configure the server identity explicitly in the GUI. It can be the same as the hostname again, i.e. `setup.vpn.cyberpointer.net` in this case, it just has to be configured explicitly.

**Reply by drcyberg** [reply]
This part works, the problem was really with `/etc/hosts` . Big thanks! I can't see the forest for the trees. :-)
In the last case, I would like to use the two certificate methods.
But the separation still does not work. :(
I use MSCHAPv2 for authentication, it works (Acces Accept), but the `eap_identity` section does not. :(
I always get the same IP address for any user, no matter how I change it.
In this case, the `setup` user should receive the IP address from the 11.11.11.0/24 domain, but it does not.
The `eap_identity` is set within windows (enable identity protection).

... `assigning virtual IP 10.10.10.1 to peer 'setup'` but why? Connection for the `setup` is `11.11.11.0/24`

**Here is the log:**

<details><summary>Details</summary>
<p>

Aug  3 08:09:53 vpn charon: 05[IKE] RADIUS authentication of 'setup' successful
Aug  3 08:09:53 vpn charon: 05[IKE] EAP method EAP_PEAP succeeded, MSK established
Aug  3 08:09:53 vpn charon: 05[ENC] generating IKE_AUTH response 11 [ EAP/SUCC ]
Aug  3 08:09:53 vpn charon: 05[NET] sending packet: from 193.188.192.136[4500] to 185.27.62.231[4500] (76 bytes)
Aug  3 08:09:53 vpn charon: 11[NET] received packet: from 185.27.62.231[4500] to 193.188.192.136[4500] (92 bytes)
Aug  3 08:09:53 vpn charon: 11[ENC] parsed IKE_AUTH request 12 [ AUTH ]
Aug  3 08:09:53 vpn charon: 11[IKE] authentication of '192.168.1.108' with EAP successful
Aug  3 08:09:53 vpn charon: 11[IKE] authentication of 'C=HU O=CyberPointer CN=vpn.cyberpointer.net' (myself) with EAP
Aug  3 08:09:53 vpn charon: 11[IKE] IKE_SA cp-access-accept-vpn[2] established between 193.188.192.136[C=HU O=CyberPointer CN=vpn.cyberpointer.net]...185.27.62.231[192.168.1.108]
Aug  3 08:09:53 vpn charon: 11[IKE] scheduling reauthentication in 3326s
Aug  3 08:09:53 vpn charon: 11[IKE] maximum IKE_SA lifetime 3506s
Aug  3 08:09:53 vpn charon: 11[IKE] peer requested virtual IP %any
Aug  3 08:09:53 vpn charon: 11[IKE] assigning virtual IP 10.10.10.1 to peer 'setup'
Aug  3 08:09:53 vpn charon: 11[IKE] peer requested virtual IP %any6
Aug  3 08:09:53 vpn charon: 11[IKE] no virtual IP found for %any6 requested by 'setup'
Aug  3 08:09:53 vpn charon: 11[IKE] CHILD_SA cp-access-accept-vpn{2} established with SPIs cd7fe14a_i 68f93180_o and TS 0.0.0.0/0 === 10.10.10.0/30
Aug  3 08:09:53 vpn vpn: + 192.168.1.108 10.10.10.0/30 == 185.27.62.231 -- 193.188.192.136 == 0.0.0.0/0
Aug  3 08:09:54 vpn charon: 11[ENC] generating IKE_AUTH response 12 [ AUTH CPRP(ADDR DNS DNS DNS DNS DNS DNS DNS DNS DNS DNS DNS DNS) SA TSi TSr N(AUTH_LFT) ]
Aug  3 08:09:54 vpn charon: 11[NET] sending packet: from 193.188.192.136[4500] to 185.27.62.231[4500] (316 bytes)
Aug  3 08:09:57 vpn charon: 06[NET] received packet: from 185.27.62.231[4500] to 193.188.192.136[4500] (76 bytes)
Aug  3 08:09:57 vpn charon: 06[ENC] parsed INFORMATIONAL request 13 [ D ]
Aug  3 08:09:57 vpn charon: 06[IKE] received DELETE for ESP CHILD_SA with SPI 68f93180
Aug  3 08:09:57 vpn charon: 06[IKE] closing CHILD_SA cp-access-accept-vpn{2} with SPIs cd7fe14a_i (2862 bytes) 68f93180_o (944 bytes) and TS 0.0.0.0/0 === 10.10.10.0/30
Aug  3 08:09:57 vpn charon: 06[IKE] sending DELETE for ESP CHILD_SA with SPI cd7fe14a
Aug  3 08:09:57 vpn charon: 06[IKE] CHILD_SA closed
Aug  3 08:09:57 vpn vpn: - 192.168.1.108 10.10.10.0/30 == 185.27.62.231 -- 193.188.192.136 == 0.0.0.0/0
Aug  3 08:09:57 vpn charon: 06[ENC] generating INFORMATIONAL response 13 [ D ]
Aug  3 08:09:57 vpn charon: 06[NET] sending packet: from 193.188.192.136[4500] to 185.27.62.231[4500] (76 bytes)
Aug  3 08:09:57 vpn charon: 07[NET] received packet: from 185.27.62.231[4500] to 193.188.192.136[4500] (76 bytes)
Aug  3 08:09:57 vpn charon: 07[ENC] parsed INFORMATIONAL request 14 [ D ]
Aug  3 08:09:57 vpn charon: 07[IKE] received DELETE for IKE_SA cp-access-accept-vpn[2]
Aug  3 08:09:57 vpn charon: 07[IKE] deleting IKE_SA cp-access-accept-vpn[2] between 193.188.192.136[C=HU O=CyberPointer CN=vpn.cyberpointer.net]...185.27.62.231[192.168.1.108]
Aug  3 08:09:57 vpn charon: 07[IKE] IKE_SA deleted
Aug  3 08:09:57 vpn charon: 07[ENC] generating INFORMATIONAL response 14 [ ]
Aug  3 08:09:57 vpn charon: 07[NET] sending packet: from 193.188.192.136[4500] to 185.27.62.231[4500] (76 bytes)

</p>
</details>

**Here is the ipsec.conf:**

<details><summary>Details</summary>
<p>

config setup
    charondebug="ike 1, knl 1, cfg 0"
    uniqueids=no

conn %default
    ikelifetime=60m
    keylife=20m
    rekeymargin=3m
    keyingtries=1
    keyexchange=ikev2

conn cp-vpn
    leftsendcert=always
    left=xxx.xxx.xxx.xxx
    leftauth=pubkey
    leftfirewall=yes
    leftsourceip=%config
    leftcert=server-vpn-cert.pem
    rightsendcert=never
    rightauth=eap-radius
    mobike=no
    leftauth=pubkey
    type=tunnel
    forceencaps=yes
    compress=no
    fragmentation=no
    rightid=%any
    right=%any
    ike=chacha20poly1305-sha512-curve25519-prfsha512,aes256gcm16-sha384-prfsha384-ecp384,aes256-sha1-modp1024,aes128-sha1-modp1024,aes256-aes128-sha256-sha1-modp3072-modp2048-modp1024,3des-sha1-modp1024
    esp=chacha20poly1305-sha512,aes256gcm16-ecp384,aes256-sha256,aes256-sha1-ecp256,aes256-sha1,3des-sha1

conn cp-access-accept-vpn
    also=cp-vpn
    eap_identity=%any
    leftsubnet=0.0.0.0/0
    leftid=*@vpn.cyberpointer.net
    rightsourceip=10.10.10.0/30
    rightsubnet=10.10.10.0/30
    rightdns=8.8.8.8,8.8.4.4
    auto=add

conn cp-setup-accept-vpn
    also=cp-vpn
    eap_identity=setup
    leftid=*@setup.vpn.cyberpointer.net
    leftsubnet=11.11.11.0/24
    rightsourceip=11.11.11.0/24
    rightsubnet=11.11.11.0/24
    auto=add
    inactivity=10

</p>
</details>

What do you think could be the problem, where am I going wrong? :(

Thank you!

**Reply by tobiasbrunner** [reply, marked as answer]
Connection switching based on EAP-Identity only works with a workaround. See e.g. my answer in #1326.

**Reply by drcyberg** [reply]
Thank you for all the help and for solving it so quickly. By the way, I was just looking at this page. :-)
ps.: _Now i am fascinated by freeradius, i suffer with groups._

Regards!

**Reply by drcyberg** [reply]
With great difficulty, but finally managed to make it work. :-)
Among other things, as you [wrote](https://docs.strongswan.org/docs/5.9/plugins/eap-radius.html#_group_selection), the solution was made with the help of the [CLASS attribute](https://freeradius.org/rfc/rfc2865.html#Class) within FreeRADIUS.

p.s.: The debugging solution helps a lot.

Thank you @tobiasbrunner !
